File: src/service/ExpenseService.ts
import { Expense, CreateExpenseRequest, ExpenseQueryParams } from '../model/Expense';
import { ExpenseRepository } from '../repository/ExpenseRepository';
import { v4 as uuidv4 } from 'uuid';

export class ExpenseService {
  private repository: ExpenseRepository;

  constructor() {
    this.repository = new ExpenseRepository();
  }

  async createExpense(request: CreateExpenseRequest): Promise<Expense> {
    // Basic validation
    if (request.amount <= 0) {
      throw new Error('Amount must be greater than 0');
    }
    
    if (!request.category || !request.description || !request.date) {
      throw new Error('Category, description, and date are required');
    }

    const expense: Expense = {
      id: uuidv4(),
      amount: request.amount,
      category: request.category,
      description: request.description,
      date: request.date,
      created_at: new Date().toISOString()
    };

    return await this.repository.save(expense);
  }

  async getExpenses(queryParams: ExpenseQueryParams = {}): Promise<Expense[]> {
    let expenses = await this.repository.findAll();

    // Filter by category if provided
    if (queryParams.category) {
      expenses = expenses.filter(e => 
        e.category.toLowerCase() === queryParams.category!.toLowerCase()
      );
    }

    // Sort by date (newest first) if sort=date_desc
    if (queryParams.sort === 'date_desc') {
      expenses.sort((a, b) => {
        const dateA = new Date(a.date).getTime();
        const dateB = new Date(b.date).getTime();
        return dateB - dateA; // Descending order
      });
    } else {
      // Default: sort by created_at descending (newest first)
      expenses.sort((a, b) => {
        const dateA = new Date(a.created_at).getTime();
        const dateB = new Date(b.created_at).getTime();
        return dateB - dateA;
      });
    }

    return expenses;
  }

  calculateTotal(expenses: Expense[]): number {
    return expenses.reduce((sum, expense) => sum + expense.amount, 0);
  }
}

File: backend/src/main/java/com/expensetracker/service/ExpenseService.java
package com.expensetracker.service;

import com.expensetracker.model.CreateExpenseRequest;
import com.expensetracker.model.Expense;
import com.expensetracker.repository.ExpenseRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class ExpenseService {
    private final ExpenseRepository repository;

    public ExpenseService(ExpenseRepository repository) {
        this.repository = repository;
    }

    public Expense createExpense(CreateExpenseRequest request) {
        // Validation is handled by @Valid annotation in controller
        Expense expense = new Expense();
        expense.setId(UUID.randomUUID().toString());
        expense.setAmount(request.getAmount());
        expense.setCategory(request.getCategory());
        expense.setDescription(request.getDescription());
        expense.setDate(request.getDate());
        expense.setCreatedAt(LocalDateTime.now());

        return repository.save(expense);
    }

    public List<Expense> getExpenses(String category, String sort) {
        List<Expense> expenses = repository.findAll();

        // Filter by category if provided
        if (category != null && !category.isEmpty()) {
            expenses = expenses.stream()
                .filter(e -> e.getCategory().equalsIgnoreCase(category))
                .collect(Collectors.toList());
        }

        // Sort by date
        if ("date_desc".equals(sort)) {
            expenses.sort(Comparator.comparing(Expense::getDate).reversed());
        } else {
            // Default: sort by created_at descending (newest first)
            expenses.sort(Comparator.comparing(Expense::getCreatedAt).reversed());
        }

        return expenses;
    }

    public Double calculateTotal(List<Expense> expenses) {
        return expenses.stream()
            .mapToDouble(Expense::getAmount)
            .sum();
    }
}

